File: templates/problems/07_trees/04_balanced_tree/balanced_tree_test.go
package balanced_tree

import "testing"

func TestIsBalanced(t *testing.T) {
	tests := []struct {
		name string
		vals []int
		want bool
	}{
		{"balanced", []int{3, 9, 20, -101, -101, 15, 7}, true},
		{"unbalanced", []int{1, 2, 2, 3, 3, -101, -101, 4, 4}, false},
		{"empty tree", []int{}, true},
		{"single node", []int{1}, true},
		{"two levels balanced", []int{1, 2, 3}, true},
		{"left heavy by one", []int{1, 2, -101, 3}, false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			root := buildTreeHelper(tt.vals)
			if got := isBalanced(root); got != tt.want {
				t.Errorf("isBalanced() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: templates/problems/07_trees/04_balanced_tree/typescript/balanced_tree.test.ts
import { describe, it, expect } from "vitest";
import { buildTree } from "./helpers";
import { isBalanced } from "./balanced_tree";

describe("isBalanced", () => {
    it.each([
        { name: "balanced", vals: [3, 9, 20, -101, -101, 15, 7], want: true },
        { name: "unbalanced", vals: [1, 2, 2, 3, 3, -101, -101, 4, 4], want: false },
        { name: "empty tree", vals: [], want: true },
        { name: "single node", vals: [1], want: true },
        { name: "two levels balanced", vals: [1, 2, 3], want: true },
        { name: "left heavy by one", vals: [1, 2, -101, 3], want: false },
        { name: "deep unbalanced with right depth 0", vals: [1, 2, -101, 3, -101, 4], want: false },
    ])("case: $name", ({ vals, want }) => {
        const root = buildTree(vals);
        expect(isBalanced(root)).toBe(want);
    });
});
